### PR TITLE
Use optnet and clearState

### PR DESCRIPTION
**Pull request by szagoruyko**
This is not ready for merge yet, but it's working and should be very useful for those who need small checkpoints or low on GPU memory. I've been using it for a while.

I removed the manual memory optimization and used optnet which does it automatically, so what this PR does is:
1. optimize memory with optnet, train
2. remove optnet optimizations
3. call clearState, save
go to step 1.

### CODE REVIEW

**Comment by facebook-github-bot-4** [comment]
Thank you for your pull request.  As you may know, we require contributors to sign our Contributor License Agreement, and we don't seem to have you on file and listed as active anymore.  In order for us to review and merge your code, please email cla@fb.com with your details so we can update your status.

**Comment by szagoruyko** [comment]
Updated the code to not use any additional memory after the first epoch. Is done by recreating model structure in new tables keeping tensor references and calling float+clearState on this new structure, thus keeping the original network untouched all the time. Checked that gradients are the same with optnet and without just in case, and that checkpointing works well.
@colesbury I think we can merge it.

**Comment by colesbury** [comment]
Thanks @szagoruyko!

This seems like the best way of doing checkpointing. 

I've moved optnet behind an `-optnet true` flag. Some large networks still don't fit with optnet. For example, ResNet-101 (batch size 256) fits on 4 GPUs with shareGradInput, but not with optnet.

Optnet seems much more robust to changes in model definition. I think I'll update the examples in the documentation to use `-optnet true` instead of `-shareGradInput true`.

**Comment by fmassa** [comment]
@colesbury I'll have a look at this difference in `optnet` this weekend. I think that this pending PR https://github.com/fmassa/optimize-net/pull/5 should remove the differences between `optnet` and `shareGradInput`, but I should double check it before merging.

**Comment by fmassa** [comment]
@colesbury @szagoruyko I just checked ResNet-101 but applying the pending PR in `optnet`.
On 1 GPU with batch size of 64, the memory consumption for a forward/barckward pass is of slightly less than 8GB (via `nvidia-smi`), whereas with the current version of `optnet` it indeed doesn't fit in memory.
I haven't tried running the full training code though, but I think 4GB might be enough for the extra memory requirements from `DataParallelTable`.

**Comment by fmassa** [comment]
For info, https://github.com/fmassa/optimize-net/pull/5 was merged into master.
@szagoruyko has been using this branch for some time and it seems to work well, and ResNet-101 fits in 4 GPUs with batch size 256.
